controllers/buildpipeline/buildpipeline_adapter.go

```diff
@@ -129,6 +133,30 @@ func (a *Adapter) EnsureSnapshotExists() (controller.OperationResult, error) {
 		return controller.RequeueWithError(err)
 	}
 
+	return a.removeFinalizerAndContinueProcessing()
+}
+
+func (a *Adapter) EnsurePipelineIsFinalized() (controller.OperationResult, error) {
+	// Is this necessary in a function that runs only when the pipeline is created?
```

Comment:
This function also runs when the pipelineRun succeeds on account of the [existing predicate](https://github.com/redhat-appstudio/integration-service/blob/d2b50365a889d09028b6324ee8c958aa818889b3/controllers/buildpipeline/buildpipeline_controller.go#L128).
Additionally, all Ensure* functions need to be idempotent because all CRs are automatically reconciled on service start up without regard for predicate filters.